File: backend/python-api/app/service/oauth_service.py
from __future__ import annotations

import logging
import secrets
from dataclasses import dataclass

import httpx
from passlib.hash import bcrypt

from app.domain.errors import ERR_INVALID_INPUT, ERR_UNAUTHORIZED, DomainError, TwoFARequiredError
from app.domain.login_event import LoginEvent
from app.domain.user import User
from app.dto.auth import TokenPair
from app.platform.geoip import GeoIPResolver
from app.repository.login_event_repo import LoginEventRepository
from app.repository.user_repo import UserRepository
from app.service.token_service import TokenService

logger = logging.getLogger(__name__)


@dataclass
class GoogleLoginResult:
    user: User
    token_pair: TokenPair | None
    requires_2fa: bool = False
    two_fa_type: str = ""


class OAuthService:
    def __init__(
        self,
        users: UserRepository,
        login_events: LoginEventRepository,
        token_svc: TokenService,
        geoip: GeoIPResolver,
        google_client_id: str,
    ):
        self._users = users
        self._login_events = login_events
        self._token_svc = token_svc
        self._geoip = geoip
        self._google_client_id = google_client_id

    async def google_login(
        self, id_token: str, remember_me: bool, client_ip: str, user_agent: str, session
    ) -> GoogleLoginResult:
        if not self._google_client_id:
            raise DomainError(ERR_INVALID_INPUT, "Google OAuth is not configured")

        payload = await self._verify_google_token(id_token)
        if payload is None:
            raise DomainError(ERR_UNAUTHORIZED, "Invalid Google authentication token")

        email = payload.get("email", "")
        if not email:
            raise DomainError(ERR_UNAUTHORIZED, "Google account email not found")

        first_name = payload.get("given_name", "")
        last_name = payload.get("family_name", "")

        existing = await self._users.find_by_email_unscoped(email)

        if existing and not existing.is_deleted:
            user = existing
            if not user.first_name and first_name:
                user.first_name = first_name
            if not user.last_name and last_name:
                user.last_name = last_name
            if not user.verified:
                user.verified = True
            await self._users.update(user)
        else:
            if existing and existing.is_deleted:
                await self._users.hard_delete_related_records(existing.id)
                await self._users.hard_delete(existing.id)

            random_pw = bcrypt.using(rounds=12).hash(secrets.token_hex(32))
            country = await self._geoip.resolve_country(client_ip)

            user = User(
                email=email,
                first_name=first_name,
                last_name=last_name,
                password_hash=random_pw,
                role="user",
                verified=True,
                language=payload.get("locale", "en")[:2].lower(),
                country=country or "",
            )
            await self._users.create(user)

        if user.two_fa_enabled:
            await session.commit()
            return GoogleLoginResult(user=user, token_pair=None, requires_2fa=True, two_fa_type="totp")

        pair = await self._token_svc.issue_token_pair(user.id, user.role, remember_me)
        event = LoginEvent(user_id=user.id, ip=client_ip, user_agent=user_agent)
        await self._login_events.create(event)
        await session.commit()

        return GoogleLoginResult(user=user, token_pair=pair)

    async def _verify_google_token(self, id_token: str) -> dict | None:
        try:
            async with httpx.AsyncClient() as client:
                resp = await client.get(
                    f"https://oauth2.googleapis.com/tokeninfo?id_token={id_token}"
                )
                if resp.status_code != 200:
                    return None
                data = resp.json()
                if data.get("aud") != self._google_client_id:
                    return None
                return data
        except Exception:
            logger.exception("Google token verification failed")
            return None

File: templates/backends/go-gin/internal/service/oauth_service.go
package service

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"log"
	"strings"
	"time"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/platform/geoip"
	"saas-starter/backend/go-api/internal/repository"
	"saas-starter/backend/go-api/pkg/i18n"

	"golang.org/x/crypto/bcrypt"
	"google.golang.org/api/idtoken"
)

type OAuthService struct {
	users       repository.UserRepository
	loginEvents repository.LoginEventRepository
	tokenSvc    *TokenService
	geoIP       geoip.Resolver
	googleClientID string
}

func NewOAuthService(
	users repository.UserRepository,
	loginEvents repository.LoginEventRepository,
	tokenSvc *TokenService,
	geoIPResolver geoip.Resolver,
	googleClientID string,
) *OAuthService {
	return &OAuthService{
		users: users, loginEvents: loginEvents,
		tokenSvc: tokenSvc, geoIP: geoIPResolver,
		googleClientID: googleClientID,
	}
}

type GoogleLoginResult struct {
	User       *domain.User
	TokenPair  *dto.TokenPair
	Requires2FA bool
	TwoFAType  string
}

func (s *OAuthService) GoogleLogin(ctx context.Context, googleToken string, rememberMe bool, clientIP, userAgent string) (*GoogleLoginResult, error) {
	if s.googleClientID == "" {
		return nil, domain.NewError(domain.ErrInvalidInput, "Google OAuth not configured")
	}

	payload, err := idtoken.Validate(ctx, googleToken, s.googleClientID)
	if err != nil {
		return nil, domain.NewError(domain.ErrUnauthorized, "invalid Google token")
	}

	email, ok := payload.Claims["email"].(string)
	if !ok || email == "" {
		return nil, domain.NewError(domain.ErrInvalidInput, "Google email not found")
	}

	firstName, lastName := extractGoogleName(payload.Claims)

	user, err := s.users.FindByEmail(ctx, email)
	if errors.Is(err, domain.ErrNotFound) {
		user, err = s.createGoogleUser(ctx, email, firstName, lastName, payload.Claims, clientIP)
		if err != nil {
			return nil, err
		}
	} else if err != nil {
		return nil, err
	} else {
		s.updateExistingGoogleUser(ctx, user, firstName, lastName)
	}

	if user.TwoFAEnabled && user.TwoFASecret != "" {
		return &GoogleLoginResult{
			User:        user,
			Requires2FA: true,
			TwoFAType:   "totp",
		}, nil
	}

	pair, err := s.tokenSvc.IssueTokenPair(ctx, user.ID, user.Role, rememberMe)
	if err != nil {
		return nil, err
	}

	go s.recordLogin(user.ID, clientIP, userAgent)

	return &GoogleLoginResult{User: user, TokenPair: pair}, nil
}

func (s *OAuthService) createGoogleUser(ctx context.Context, email, firstName, lastName string, claims map[string]interface{}, clientIP string) (*domain.User, error) {
	randomBytes := make([]byte, 32)
	if _, err := rand.Read(randomBytes); err != nil {
		return nil, err
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(hex.EncodeToString(randomBytes)), BcryptCost)
	if err != nil {
		return nil, err
	}

	userLang := "en"
	if locale, ok := claims["locale"].(string); ok && locale != "" {
		parts := strings.Split(locale, "_")
		if len(parts) > 0 {
			code := strings.ToLower(parts[0])
			if i18n.IsLanguageSupported(code) {
				userLang = code
			}
		}
	}

	country := s.geoIP.CountryFromIP(clientIP)

	user := &domain.User{
		FirstName:    firstName,
		LastName:     lastName,
		Email:        email,
		PasswordHash: string(hash),
		Role:         "user",
		Verified:     true,
		Language:     userLang,
		Country:      country,
	}

	if err := s.users.Create(ctx, user); err != nil {
		return nil, err
	}

	return user, nil
}

func (s *OAuthService) updateExistingGoogleUser(ctx context.Context, user *domain.User, firstName, lastName string) {
	updated := false
	if user.FirstName == "" && firstName != "" {
		user.FirstName = firstName
		updated = true
	}
	if user.LastName == "" && lastName != "" {
		user.LastName = lastName
		updated = true
	}
	if !user.Verified {
		user.Verified = true
		updated = true
	}
	if updated {
		_ = s.users.Update(ctx, user)
	}
}

func (s *OAuthService) recordLogin(userID uint, ip, userAgent string) {
	event := &domain.LoginEvent{
		UserID: userID, IP: ip, UserAgent: userAgent, LoggedAt: time.Now(),
	}
	if err := s.loginEvents.Create(context.Background(), event); err != nil {
		log.Printf("Failed to record login event for user %d: %v", userID, err)
	}
}

func extractGoogleName(claims map[string]interface{}) (string, string) {
	firstName, lastName := "", ""

	if name, ok := claims["name"].(string); ok && name != "" {
		parts := strings.Fields(name)
		if len(parts) > 0 {
			firstName = parts[0]
		}
		if len(parts) > 1 {
			lastName = strings.Join(parts[1:], " ")
		}
	} else {
		if given, ok := claims["given_name"].(string); ok {
			firstName = given
		}
		if family, ok := claims["family_name"].(string); ok {
			lastName = family
		}
	}

	return firstName, lastName
}
